neo/Consensus/ConsensusService.cs

```diff
@@ -287,19 +333,154 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
+                context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            {
+                Block block = context.MakeHeader();
+                if (block == null) return;
+                context.State |= ConsensusState.CommitSent;
+                context.FinalSignatures[context.MyIndex] = block.Sign(context.KeyPair);
+                SignAndRelay(context.MakeCommitAgreement(context.FinalSignatures[context.MyIndex]));
+                Log($"Commit sent: height={context.BlockIndex} hash={block.Hash} state={context.State}");
+            }
+        }
+
+
+        private void OnCommitAgreement(ConsensusPayload payload, CommitAgreement message)
+        {
+            if (context.FinalSignatures[payload.ValidatorIndex] != null) return;
+
+            Log($"{nameof(OnCommitAgreement)}: height={payload.BlockIndex} hash={context.MakeHeader().Hash.ToString()} view={message.ViewNumber} index={payload.ValidatorIndex}");
+
+            if (!Crypto.Default.VerifySignature(context.MakeHeader().GetHashData(), message.FinalSignature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+            {
+                Log($"{nameof(OnCommitAgreement)}: SIGNATURE verification with problem");
+                return;
+            }
+
+            context.FinalSignatures[payload.ValidatorIndex] = message.FinalSignature;
+
+            if (context.FinalSignatures.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            {
+                Contract contract = Contract.CreateMultiSigContract(context.M, context.Validators);
+                Block block = context.MakeHeader();
+                if (block == null) return;
+                context.State |= ConsensusState.BlockSent;
+
+                ContractParametersContext sc = new ContractParametersContext(block);
+                for (int i = 0, j = 0; i < context.Validators.Length && j < context.M; i++)
+                    if (context.FinalSignatures[i] != null)
+                    {
+                        sc.AddSignature(contract, context.Validators[i], context.FinalSignatures[i]);
+                        j++;
+                    }
+                sc.Verifiable.Witnesses = sc.GetWitnesses();
+                block.Transactions = context.TransactionHashes.Select(p => context.Transactions[p]).ToArray();
+                Log($"relay block: height={context.BlockIndex} hash={block.Hash}");
+
+                system.LocalNode.Tell(new LocalNode.Relay { Inventory = block });
+            }
+        }
+
+        private void OnRenegeration(ConsensusPayload payload, Renegeration message)
+        {
+            Log($"{nameof(OnRenegeration)}: height={payload.BlockIndex} hash={context.MakeHeader().Hash.ToString()} view={message.ViewNumber} numberOfPartialSignatures={message.SignedPayloads.Count(p => p != null)} index={payload.ValidatorIndex}");
```

Comment:
when a CN is in commit phase, it should ignore `renegeration message` from other CN which also in commit phase. right ?

BTW, is renegeration a typo? Also feel so wrong typing it. Should it be **regeneration**? just wondering.

Reply:
Yes, right.

It is blocked by `OnConsensusPayload`, it has a flag for `CommitSent`

Fell free to suggest another name for this stage, regeneration is like synonymous of recovery, recuperation, regeneration, retrieval, reclamation, repair

Reply:
Sorry, @wanglongfei88!

You are right again,
I just got this error. The flag there is `BlockSent`